### DISCUSSION TITLE
Enabling zipped downloads from folders or bulk file selections (RFC for FR #12980)

### DISCUSSION THREAD

**Discussion by dafrica1** [directus/directus#20601] [Feature Requests]
### Summary

_This is a continuation of [a Feature Request submitted in  #2020](https://github.com/directus/directus/discussions/12980) by @benhaynes which has already been moved forward to "Under Review" but requires an RFC. As this is something I'm currently facing, I thought I'd start the process by creating the RFC and contributing to the previous discussion._

The intention of this FR is to allow users to download multiple documents at a time from folders or by manually selecting multiple files in the form of a ZIP file which can be downloaded locally. This would improve UX for file/asset heavy workflows and bring Directus capabilities in line with similar platforms and CMSs.

### Motivation

Directus has a very attractive file module but it currently lacks some common functionality that would make Directus a top choice for use cases that involve document management or file-based assets. The File module already supports the concept of bulk operations and export functionality is already available for database-managed content but the same consideration has not been given to files and folders at this point.

Bulk operations are a key component of any file-based workflow and, combined with another FR (#20600), would address some low hanging fruit that would unlock additional usecases for Directus customers.

### Detailed Design

![Screenshot from 2023-12-01 09-05-03](https://github.com/directus/directus/assets/44697531/a007a104-9482-4423-a28a-4c2c172fdc0f)
![Screenshot from 2023-12-01 09-44-12](https://github.com/directus/directus/assets/44697531/1bca4b7c-04d0-40d5-879d-3c7c1da7712a)

**Mockup: Includes a download button next to the folder creation button that could perform the following actions when clicked:**

- If no bulk selection is made by the user, auto-select all visible files (and nested folders and their contents) to be included in the list of files to download. The goal would be to maintain the nested folded hierarchy as part of the final zip file.
- Display the modal from the second mockup to confirm the user's selection and initiate the download.
- The download could kick off a background task to zip the files asynchronously and either initiate the download immediately or create a notification to the user when the download is ready.
- An audit log record could then be made of the transaction, the files included, and any information regarding versioning and such

### Must Have:
- Ability to zip multiple files based on user selection or folder contents and provide them for download
- Maintain nested folder hierarchy

### Should Have:
- Ability for user to remove or append additional files to include in the ZIP file (my mockup of the modal is lacking this but it could be a further iteration of the existing idea)
- Entry in the audit log for a download initiated, the files included, their versioning details if enabled, along with the accountability data
- Default zip file name is the folder name (if a folder is selected) otherwise `files-<timestamp>.zip`

### Could Have:
- Ability to include any metadata associated with the selected files embedded within the files themselves (e.g. Word documents having tags applied
- Ability to run as an async background task depending on the total size of the documents selected and trigger a notification within Directus or via e-mail
- Ability to display selected nested folders in a tree view with checkboxes to include/exclude specific nested files
- Download actions as triggers for Flows

### Won't Have:
- Ability to change compression type

### Drawbacks

- Depending on the size of the selected files and the system resources, this could be a resource intensive process
- Inclusion of metadata within the files themselves may vary in viability due to differences between file formats
- May be complications in implementation depending on the storage driver used
- Unclear how performant the process of recursively capturing nested files/folders info is within Directus

### Alternatives

- From a UX perspective, the modal is one potential option. Others include incorporating a download function within the right sidebar in the same location as Import/Export functionality for collection items.
- A flyout menu on the side could be leveraged instead of a modal

### Adoption Strategy

I don't see any adoption issues as this wouldn't interfere with any other functionality or break existing capabilities. The intent would be to bring basic feature parity with other file explorers into Directus.

**Comment by benhaynes** [comment]
Thank you!!! Curious... would it be simpler to just use the File Library _select_ functionality, and then add a "Download" action button to the header. If you have <X items selected it downloads the individual assets... otherwise it zips and downloads with a default naming convention (`project-name-download-231203A` timestamp?).

Just trying to keep this as simple as possible. 🙌

**Reply by dafrica1** [reply]
For an individual file, having the download action button would definitely work and for more than one file, the zip option might be the way to go as I haven't seen a better way to download multiple distinct files in one go yet.

The simpler the better!

**Reply by benhaynes** [reply]
True! Anything over one file could use the `zip` approach... but the simple header button might be cleaner than a modal. The only thing that might not have is including files to download within different nested folders.

**Reply by dafrica1** [reply]
Agreed, header button would be better. RE: modification to the selection - as I think on it more, I think that might be unnecessary as well.

Selecting distinct items in nested folders isn't common in any file explorer I've seen so it can probably be done away with and someone can create an extension for such a function if they really need it.

So end result: No modal, just a header button that checks to see if more than one file is selected.
- If only one file is selected, download it as an individual file
- If more than one file is selected, turn it into an async background job (hard to predict how quickly the storage adapter will respond so probably best not to leave the user waiting), then trigger a notification when the zip is ready for download (might need some temp storage that autodeletes the generated file after 24hrs or some set period of time to prevent storage use from growing unexpectedly)

**Reply by yisheng** [reply]
> If more than one file is selected, turn it into an async background job (hard to predict how quickly the storage adapter will respond so probably best not to leave the user waiting), then trigger a notification when the zip is ready for download (might need some temp storage that autodeletes the generated file after 24hrs or some set period of time to prevent storage use from growing unexpectedly)

Another option is to zip file using [jszip](https://github.com/Stuk/jszip), no server-side background job needed.

**Comment by w0kyj** [comment]
Linked as RFC for https://github.com/directus/directus/discussions/12980

**Comment by lucameusburger** [comment]
This feature is still not implemented as far as i know. For everyone landing on this thread: i made a custom endpoint that can zip a folder (including subfolders and files recursively) and downloads it. And additionally a POST endpoint where you can pass file-ids that then also get zipped and downloaded. while building this i was thinking a lot about file-caching and how to best handle this. In a proper implementation we would probably have some directus env settings like DL_FILE_CACHE_ENABLED and DL_FILE_CACHE_TTL or something. Anyway here is the code:

```
import archiver from 'archiver';
// Types are used in interface definitions
import { defineEndpoint } from '@directus/extensions-sdk';

interface FileItem {
	id: string;
	filename_download: string;
	type?: string | null;
	folder?: string | null;
	title?: string | null;
}

interface FolderItem {
	id: string;
	name: string;
	parent?: string | null;
}

export default defineEndpoint((router, { services, getSchema }) => {
	const { AssetsService, FilesService, FoldersService } = services;

	/**
	 * Recursively collect all files and subfolders from a parent folder
	 */
	async function collectFolderContentsRecursive(
		folderId: string,
		schema: any,
		accountability: any,
		path: string = ''
	): Promise<{ files: FileItem[], folders: FolderItem[], paths: Map<string, string> }> {
		const filesService = new FilesService({ schema, accountability });
		const foldersService = new FoldersService({ schema, accountability });
		
		const allFiles: FileItem[] = [];
		const allFolders: FolderItem[] = [];
		const filePaths = new Map<string, string>();

		try {
			// Get all files in current folder
			const files = await filesService.readByQuery({
				filter: { folder: { _eq: folderId } },
				fields: ['id', 'filename_download', 'type', 'folder', 'title'],
				limit: -1
			});

			// Add files to collection with their paths
			for (const file of files) {
				allFiles.push(file);
				const fileName = file.filename_download || file.title || `file_${file.id}`;
				filePaths.set(file.id, path ? `${path}/${fileName}` : fileName);
			}

			// Get all subfolders in current folder
			const folders = await foldersService.readByQuery({
				filter: { parent: { _eq: folderId } },
				fields: ['id', 'name', 'parent'],
				limit: -1
			});

			// Recursively process each subfolder
			for (const folder of folders) {
				allFolders.push(folder);
				const folderPath = path ? `${path}/${folder.name}` : folder.name;
				
				const subContents = await collectFolderContentsRecursive(
					folder.id,
					schema,
					accountability,
					folderPath
				);
				
				allFiles.push(...subContents.files);
				allFolders.push(...subContents.folders);
				
				// Merge path maps
				subContents.paths.forEach((value, key) => {
					filePaths.set(key, value);
				});
			}
		} catch (error) {
			console.error(`Error collecting folder contents for folder ${folderId}:`, error);
			throw error;
		}

		return { files: allFiles, folders: allFolders, paths: filePaths };
	}

	/**
	 * Create and stream a ZIP archive containing the provided files
	 */
	async function createZipStream(
		files: FileItem[],
		filePaths: Map<string, string>,
		assetsService: any,
		res: any
	): Promise<void> {
		return new Promise((resolve, reject) => {
			const archive = archiver('zip', {
				zlib: { level: 9 } // Maximum compression
			});

			// Set response headers for ZIP download
			res.setHeader('Content-Type', 'application/zip');
			res.setHeader('Content-Disposition', 'attachment; filename="download.zip"');

			// Pipe archive to response
			archive.pipe(res);

			// Handle archive events
			archive.on('error', (err: Error) => {
				console.error('Archive error:', err);
				reject(err);
			});

			archive.on('end', () => {
				console.log('Archive finalized');
				resolve();
			});

			// Add files to archive
			const addFilesToArchive = async () => {
				try {
					for (const file of files) {
						const filePath = filePaths.get(file.id) || file.filename_download || `file_${file.id}`;
						
						try {
							// Get file stream from AssetsService
							const { stream } = await assetsService.getAsset(file.id);
							
							if (stream) {
								archive.append(stream, { name: filePath });
							} else {
								console.warn(`No stream available for file ${file.id}`);
							}
						} catch (fileError) {
							console.error(`Error adding file ${file.id} to archive:`, fileError);
							// Continue with other files instead of failing completely
						}
					}

					// Finalize the archive
					await archive.finalize();
				} catch (error) {
					reject(error);
				}
			};

			addFilesToArchive();
		});
	}

	// Download folder and all nested content as ZIP
	router.get('/folder/:id', async (req: any, res: any) => {
		try {
			const folderId = req.params.id;
			console.log('Bulk download requested for folder:', folderId);

			const schema = await getSchema();
			const assetsService = new AssetsService({ 
				schema, 
				accountability: req.accountability 
			});
			const foldersService = new FoldersService({ 
				schema, 
				accountability: req.accountability 
			});

			// Verify folder exists
			const folder = await foldersService.readOne(folderId, {
				fields: ['id', 'name']
			});

			if (!folder) {
				return res.status(404).json({
					error: 'Folder not found'
				});
			}

			// Collect all files and folders recursively
			const { files, paths } = await collectFolderContentsRecursive(
				folderId,
				schema,
				req.accountability
			);

			console.log(`Found ${files.length} files to download from folder: ${folder.name}`);

			if (files.length === 0) {
				return res.status(400).json({
					error: 'No files found in the specified folder'
				});
			}

			// Set filename based on folder name
			const zipFilename = `${folder.name.replace(/[^a-zA-Z0-9]/g, '_')}.zip`;
			res.setHeader('Content-Disposition', `attachment; filename="${zipFilename}"`);

			// Create and stream ZIP
			await createZipStream(files, paths, assetsService, res);

		} catch (error) {
			console.error('Error in bulk folder download:', error);
			
			if (!res.headersSent) {
				return res.status(500).json({
					error: 'Failed to create download archive',
					details: error instanceof Error ? error.message : 'Unknown error'
				});
			}
		}
	});

	// Download multiple files as ZIP
	router.post('/files', async (req: any, res: any) => {
		try {
			const { fileIds } = req.body;
			
			if (!fileIds || !Array.isArray(fileIds) || fileIds.length === 0) {
				return res.status(400).json({
					error: 'fileIds array is required and must not be empty'
				});
			}

			console.log('Bulk download requested for files:', fileIds);

			const schema = await getSchema();
			const filesService = new FilesService({ 
				schema, 
				accountability: req.accountability 
			});
			const assetsService = new AssetsService({ 
				schema, 
				accountability: req.accountability 
			});

			// Get file information
			const files = await filesService.readByQuery({
				filter: { id: { _in: fileIds } },
				fields: ['id', 'filename_download', 'type', 'title'],
				limit: -1
			});

			if (files.length === 0) {
				return res.status(404).json({
					error: 'No files found with the provided IDs'
				});
			}

			// Create simple path mapping for individual files
			const filePaths = new Map<string, string>();
			files.forEach((file: FileItem) => {
				const fileName = file.filename_download || file.title || `file_${file.id}`;
				filePaths.set(file.id, fileName);
			});

			console.log(`Found ${files.length} files to download`);

			// Set ZIP filename
			const zipFilename = `files_${new Date().toISOString().split('T')[0]}.zip`;
			res.setHeader('Content-Disposition', `attachment; filename="${zipFilename}"`);

			// Create and stream ZIP
			await createZipStream(files, filePaths, assetsService, res);

		} catch (error) {
			console.error('Error in bulk files download:', error);
			
			if (!res.headersSent) {
				return res.status(500).json({
					error: 'Failed to create download archive',
					details: error instanceof Error ? error.message : 'Unknown error'
				});
			}
		}
	});
});
```

**Comment by zguig52** [comment]
Just did a custom endpoint extension based on your code @lucameusburger, so it is easy to integrate. Thanks a lot for sharing your code!
https://www.npmjs.com/package/directus-extension-downloadfolderorfilesaszip-endpoint

**Comment by JamesW1** [comment]
**Update: Bulk file and folder downloads are now available in v11.14 🎉**
We've shipped the ability to download multiple files and entire folders as ZIP archives. 

Docs here:  https://directus.io/docs/guides/files/download

Thanks @dafrica, @lucameusburger and everyone who contributed to the discussion and built workarounds until we got round to this!

**Reply by dafrica1** [reply]
Thank you @JamesW1 and the Directus team for making this happen. Much appreciated!
